src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/Unix/DecryptorPalOpenSsl.DecodeRecipients.cs

```diff
@@ -0,0 +1,306 @@
+
+            List<KeyAgreeRecipientInfo> keyAgreeRecipients = new List<KeyAgreeRecipientInfo>(recipientIdSet.Count);
+            for (int i = 0; i < recipientIdSet.Count; i++)
+            {
+                KeyAgreeRecipientInfoPalOpenSsl kari =
+                    new KeyAgreeRecipientInfoPalOpenSsl(
+                        version,
+                        recipientIdSet[i],
+                        originatorIdentifierOrKey,
+                        otherKeyMaterialSet[i],
+                        keyEncryptionAlgo,
+                        encryptionKeySet[i],
+                        dateSet[i]);
+                keyAgreeRecipients.Add(new KeyAgreeRecipientInfo(kari));
+            }
+            return keyAgreeRecipients;
+        }
+
+        private static PublicKeyInfo ReadOriginatorPublicKey(DerSequenceReader encodedKeyAgreement)
+        {
+            // OriginatorPublicKey::= SEQUENCE {
+            //      algorithm AlgorithmIdentifier,
+            //      publicKey BIT STRING }
+            DerSequenceReader publicKeyReader = encodedKeyAgreement.ReadSequence();
+
+            DerSequenceReader algoIdentifier = publicKeyReader.ReadSequence();
+
+            Oid algoOid = algoIdentifier.ReadOid();
+            byte[] publicKey = publicKeyReader.ReadBitString();
+ 
+            // This section reads the key and determines the bit length of the key
+            byte[] rawKey = DerSequenceReader.CreateForPayload(publicKey).ReadIntegerBytes();
+            int trailingZeros = 0;
+            while (rawKey[trailingZeros++] == 0);
+            int keyLength = 8 * (rawKey.Length - trailingZeros);
+
+            AlgorithmIdentifier publicKeyAlgorithm = new AlgorithmIdentifier(algoOid, keyLength);
+
+            return new PublicKeyInfo(publicKeyAlgorithm, publicKey);
+        }
+
+        private static KeyTransRecipientInfo ReadKeyTrans(DerSequenceReader recipientSet)
+        {
+            // As of RFC 2630 this is a KeyTransRecipientInfo 
+            //
+            // KeyTransRecipientInfo::= SEQUENCE {
+            //      version CMSVersion,  --always set to 0 or 2
+            //      rid RecipientIdentifier,
+            //      keyEncryptionAlgorithm KeyEncryptionAlgorithmIdentifier,
+            //      encryptedKey EncryptedKey }
+            //
+            // EncryptedKey::= OCTET STRING
+
+            DerSequenceReader encodedKeyTransport = recipientSet.ReadSequence();
+            int version = encodedKeyTransport.ReadInteger();
+
+            // RecipientIdentifier is either SubjectKeyIdentifier or IssuerAndSerialNumber
+            //
+            // RecipientIdentifier ::= CHOICE {
+            //      issuerAndSerialNumber IssuerAndSerialNumber,
+            //      subjectKeyIdentifier[0] SubjectKeyIdentifier }
+
+            byte tag = encodedKeyTransport.PeekTag();
+            
+            SubjectIdentifier subject = tag == DerSequenceReader.ContextSpecificTagFlag ?
```

Comment:
I don't know how good we've been at future-proofing in other places, but this feels slightly wrong to me.

We can positively test for IssuerAndSerialNumber (constructed sequence), and we can positively test for SubjectKeyIdentifier (primitive context specific 0), but anything which isn't one of those two values is from the future (or incorrectly encoded).